### ISSUE TITLE
Training a single image with the network

### ISSUE THREAD

**Issue by shubhashis-k** [szagoruyko/cifar.torch#11]
I have tried out your model. I was able to reproduce the results. I tried only the CPU version. It works as you mentioned. But I can't figure out how to use it to classify a single image.

What I know so far is that, I will have to forward the model with Tensor of size [2,3,32,32]. Tensor of size [1,3,32,32] won't work because of BatchNormalization. This produces result. But incorrect ones.

I have an image. Lets say, it is "car.jpg". The dimension of this image is [3,32,32]. Now how will i have to forward this image through the model? Will I have to preprocess this test image? 

What I did was, 

```
car = image.load("car.jpg")  --note that car has the dimension [3,32,32]
dataset = torch.Tensor(2,3,32,32)
dataset[1] = car
dataset[2] = car

model = torch.load("model.net")
result = model:forward(dataset)
```

But this produces incorrect results. So, what will need to do to correctly classify a single image?

**Comment by kmul00** [comment]
Hi,

If you are not using data from `testData.data` while _evaluation_, then you need to preprocess the data as shown [here](https://github.com/szagoruyko/cifar.torch/blob/master/provider.lua#L90-L106). 

Basically you need to convert the data to **YUV** from **RGB**, normalize the **Y** channel locally and perform **zero mean unit std** global normalization on the **U** and **V** channel respectively. `testData` is already normalized as can be seen from the link I provided above.

Also I think you don't need to have dimension `[2 3 32 32]` for `BatchNormalization` to work. `[1 3 32 32]` will also work  (of course during training it doesn't make sense to have batch size of 1 and perform batch norm, but in evaluation phase it should work). Did you get any error with batch size 1 ?

It worked when I tested with a demo model.

Also don't forget to set `model:evaluate()`.

**Comment by shubhashis-k** [comment]
@kmul00, I tried what you said. First, i computed the mean_u, std_u,mean_v, std_v, then i normalized. Here is my code implementation. Notice, I stored the mean_u, std_u,mean_v, std_v statically which i computed before from trainData.(As it was written on provider.lua). 

```
mean_u = 122.95039414062    
std_u  = 62.088708246722    
mean_v = 113.86538318359    
std_v  = 66.704900292063    


car = image.load("car.jpg")
testData = torch.Tensor(2,3,32,32)
normalization = nn.SpatialContrastiveNormalization(1, image.gaussian1D(7))

testData[1] = car
testData[2] = car

for i = 1,2 do
     -- rgb -> yuv
     local rgb = testData[i]
     local yuv = image.rgb2yuv(rgb)
     -- normalize y locally:
     yuv[{1}] = normalization(yuv[{{1}}])
     testData[i] = yuv
  end

  -- normalize u globally:
  testData:select(2,2):add(-mean_u)
  testData:select(2,2):div(std_u)
  -- normalize v globally:
  testData:select(2,3):add(-mean_v)
  testData:select(2,3):div(std_v)
```

The output I get is 

```
2.9241 -1.0155 1.8673 -0.5512 1.1843 -2.3183 0.7540 -1.7609 1.4749 -1.2720
2.9241 -1.0155 1.8673 -0.5512 1.1843 -2.3183 0.7540 -1.7609 1.4749 -1.2720
```

This was an image of an automobile. So the correct class should be 2. Now from the output what i see is that, the predicted model is 1(the first one has the highest value). Also I tried with some trained images. It does not classify correctly as well. 

I also set `model:evaluate()`. The result does not change. Did you have the same behavior? 

Am i interpreting something wrong?

Also,  [1 3 32 32] dimension does not work with BatchNormalization. The error it produces is 
`...ch/torch/install/share/lua/5.1/nn/BatchNormalization.lua:67: only mini-batch supported (2D tensor), got 1D tensor instead`. But this is not my problem now.

**Comment by kmul00** [comment]
I don't see anything wrong in it really. Except one concern. 

I guess `car.jpg` is one of your external image. Check once after doing `image.load('car.jpg')` that is it loaded correctly, that is whether the pixel values lie within the range `[0, 255]`. From the _mean_ and _std_ it is clear that expected image range should be that. Sometimes while loading `jpeg` images using `image.load`, it may get scaled to `[0 1]`.

Check once using `testData[1]:max()` and `testData[1]:min()` that what is the range of your image pixel values after `testData[1] = car`.

**Comment by shubhashis-k** [comment]
@kmul00 I think you are right. car:min() returns 0 and car:max() returns 1. After normalizing testData[1]:min() becomes -2.30 and testData[1]:max() becomes 2.84. Your assumption is correct. 

What should I do in this aspect?

**Comment by kmul00** [comment]
Easiest way out is to convert it to `[0 255]` by doing `car:mul(255)`.

I don't know if there is already an available method in _image_ package that let's you load image in `[0, 255]` scale. You can check the _image_ package for that, and if you don't find any suitable solution, open an issue maybe.

But my assumption is simply multiplying by 255 will do the trick.

**Comment by shubhashis-k** [comment]
I don't know what is the cause. But when I am modifying the image by car:mul(255), but after normalizing the value of testData[1]:max and testData[1]:min is almost the same. I mean testData[1]:min() becomes -2.40 and testData[1]:max() becomes 2.84. Almost same as before. I checked the values by printing car:max(), testData[1]:max before normalization. The value is within [0-255] before normalization. But the min and max is almost unchanged after normalization.

I checked the result by forward passing. No luck :(. Checked with training data. The result is still bad.

**Comment by kmul00** [comment]
That's what _zero mean unit std_ is supposed to do. Your pixel values will lie within the range of `[-3*std, +3*std]`.

That is `[-3, 3]` in this case.

**Comment by setu1421** [comment]
We ran it in our trained model but failed to give the exact result.. It is giving maximum output to a label but which is not the desired label. It is for the cases for all the images.. What could be the problem?? Is the output which is the maximum is the our predicted label or do we have to do something with the output result??

**Comment by shubhashis-k** [comment]
@kmul00 

So, It's been late. But, I was able to find out what the problem was. It turns out, I was getting different result due to the fact that the saving mechanism in torch7 is different from what I expected. 

I wrote a piece of code to test whether my saved(loaded from cifar10) and loaded image were same. It turns out they are different. So, how do I save an image that retains its original value?

```
i1 = testData.data[2] --fetching data from CIFAR-10
image.save("1.png", i) --saving the data as image

i2 = image.load("1.png") --loading the saved image

if(i1 == i2) then --checking if image1(i1) and image2(i2) are different
print("same") 
end
```

**Comment by kmul00** [comment]
@wildfire001 Do you absolutely need to save your data in image format only ? Why don't you simply use `torch.save` and save the data as tensors ? If that doesn't cause too much problem, maybe you can do that.

I am not very sure regarding the **image** package. Maybe while saving and loading in different formats (`jpeg`, `png`, e.t.c.) modifies the values somewhat.

**Comment by shubhashis-k** [comment]
There is problem here. If I save in PNG format or in jpg format, the pixel values are different. I checked. 

So, when I forward pass the network with an image(for testing purposes) should I use png or jpg format? Also I kind of need to to save it in image format. Is there any solution for that?

BTW, thanks for helping out this quickly :+1:

**Comment by kmul00** [comment]
Here's a workaround, see if it works for you.

Apparently when you save and load images using `image.load()` or `image.save()`, it works well for values `[0, 1]`. So before saving your data, perform `data = data / 255` to scale it to `[0, 1]`.

Similarly after loading project it back to original `[0, 255]` by multiplying with 255.

**Comment by shubhashis-k** [comment]
I tested it, sorry it is not retaining its original values.

The code I used is this, is this ok?

```
image1 = image/255;

image.save("adup.png", image1)

b = image.load("adup.png")

if(b == image1) then
 print("same")
end
```

**Comment by kmul00** [comment]
I suggest you ask about this in the `image` repository, along with an example. Chances are high of getting better solution there. 

I myself also am not that adept with the `image` package, so don't want to just guess things.

**Comment by shubhashis-k** [comment]
Also, I tried to save the image(The tensor) by torch.save("filename", tensor). 

The file has become very large(around 245 MB). Is this supposed to happen?

**Comment by kmul00** [comment]
Depends on the size and type of your tensor. You can check the size and type using `tensor:size()` and `tensor:type()` respectively.

Though, 245 MB for just 1 image is absurdly high. Something is wrong.

**Comment by shubhashis-k** [comment]
I just wrote `torch.save("a", image)`. image holds the Tensor, "a" is the filename to be saved. 

Am I missing something?

**Comment by kmul00** [comment]
``` lua
data = testData.data[1]:clone()
torch.save('data.t7', data)
```

This should work.

**Comment by shubhashis-k** [comment]
@kmul00 

Yes, it works. The network behavior is as expected.

Now, lets say I want to classify an image outside of CIFAR-10 testing data. I will have to download it from internet. Now what format should I use to load that image? "PNG" or "JPG" ? As both formats differ very much.

I get that after loading I will have to normalize that image. The normalization process remains the same. 

But what format should I use for saving the image from internet to test? "PNG" or "JPG"?

**Comment by fmassa** [comment]
I just created a PR with example code on how to use a random image (JPG, PNG, etc) on cifar torch https://github.com/szagoruyko/cifar.torch/pull/13

**Comment by shubhashis-k** [comment]
thanks @fmassa. This sums everything up.
